desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+        the accuracy of the quadrature.
+    method : str
+        The method to evaluate the integral.
+        The "spline" method exactly integrates a cubic spline of the integrand.
+        The "quad" method performs a Gauss quadrature and estimates the integrand
+        by using distinct cubic splines for components in the integrand so that
+        the singularity from the division by zero near the bounce points can be
+        captured more accurately than can be represented by a polynomial.
+
+    Returns
+    -------
+    ba : callable
+        This callable method computes the bounce average G_ℓ(λ) for every
+        specified field line ℓ (constant rho and alpha), for every λ value in
+        ``lambdas``.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        ba = bounce_average(eq)
+        G = ba(name, pitch)
+
+    """
+    bi = bounce_integral(eq, rho, alpha, zeta_max, pitch, resolution, method)
+
+    def _bounce_average(name, pitch=None):
+        """Compute the bounce average of the named quantity using the spline method.
+
+        Parameters
+        ----------
+        name : ndarray
+            Name of quantity in ``data_index`` to compute the bounce average of.
+        pitch : ndarray
+            λ values to evaluate the bounce average at.
+            If None, uses the values given to the parent function.
+
+        Returns
+        -------
+        G : ndarray, shape(pitch, alpha, rho, (resolution - 1) * 2)
+            The last axis iterates through every bounce average performed
+            along that field line padded by zeros.
+
+        """
+        return safediv(bi(name, pitch), bi("1", pitch))
+
+    return _bounce_average
+
+
+def field_line_to_desc_coords(eq, rho, alpha, zeta):
+    """Get DESC grid from unique field line coordinates."""
+    r, a, z = jnp.meshgrid(rho, alpha, zeta, indexing="ij")
+    r, a, z = r.ravel(), a.ravel(), z.ravel()
+    # Map these Clebsch-Type field-line coordinates to DESC coordinates.
+    # Note that the rotational transform can be computed apriori because it is a single
+    # variable function of rho, and the coordinate mapping does not change rho. Once
+    # this is known, it is simple to compute theta_PEST from alpha. Then we transform
+    # from straight field-line coordinates to DESC coordinates with the method
+    # compute_theta_coords. This is preferred over transforming from Clebsch-Type
+    # coordinates to DESC coordinates directly with the more general method
+    # map_coordinates. That method requires an initial guess to be compatible with JIT,
+    # and generating a reasonable initial guess requires computing the rotational
+    # transform to approximate theta_PEST and the poloidal stream function anyway.
+    # TODO: map coords recently updated, so maybe just switch to that
+    lg = LinearGrid(rho=rho, M=eq.M_grid, N=eq.N_grid, NFP=eq.NFP, sym=eq.sym)
+    lg_data = eq.compute("iota", grid=lg)
```

Comment:
i think this should be computing iota and its derivatives, since the derivatives are needed to compute the derivatives of B. This might be the cause of |B| going to zero in some places.

Reply:
Depending on the qty being bounce averaged, there may also be other things that need to be computed on the full grid.

Reply:
I'll look into that. FYI, in the next line, toggling between jitable=True and jitable=False for the grid construction gives different results (10 orders of magnitude for |B| as can be tested here https://github.com/PlasmaControl/DESC/pull/854#discussion_r1505121975) for computed quantities - regardless of whether `override_grid` is toggled on.